### PR TITLE
Add method for 'tested by' CodeLens

### PR DESCRIPTION
**Pull request by panopticoncentral**
In addition to the "references" CodeLens, there's also a CodeLens that indicates the # of methods that test a particular method and whether they're passing. This needs a slightly different set of information about references, so I needed to extend the interface slightly.

@heejaechang @CyrusNajmabadi

### CODE REVIEW

**Comment by panopticoncentral** [comment]
@heejaechang @CyrusNajmabadi Ping?

File: src/Features/CSharp/Portable/CodeLens/CSharpCodeLensDisplayInfoService.cs [hunk 1]
```diff
@@ -165,6 +165,34 @@ public string GetDisplayName(SemanticModel semanticModel, SyntaxNode node)
...
 
+        public string ConstructFullName(SemanticModel semanticModel, SyntaxNode methodNode)
+        {
+            var name = semanticModel.GetDeclaredSymbol(methodNode).Name;
+            var parent = methodNode.Parent;
+            while (parent != null)
+            {
+                if (parent.IsKind(SyntaxKind.ClassDeclaration))
+                {
+                    name = semanticModel.GetDeclaredSymbol(parent).Name + (!string.IsNullOrEmpty(name) ? "+" + name : "");
```
Comment (heejaechang) [inline src/Features/CSharp/Portable/CodeLens/CSharpCodeLensDisplayInfoService.cs]: string builder?

**Comment by CyrusNajmabadi** [comment]
> There's also a CodeLens that indicates the # of methods that test a particular method and whether they're passing.

how do we gather that information?

File: src/Features/CSharp/Portable/CodeLens/CSharpCodeLensDisplayInfoService.cs [hunk 2]
```diff
@@ -165,6 +165,34 @@ public string GetDisplayName(SemanticModel semanticModel, SyntaxNode node)
             }
         }
 
+        public string ConstructFullName(SemanticModel semanticModel, SyntaxNode methodNode)
+        {
+            var name = semanticModel.GetDeclaredSymbol(methodNode).Name;
```
Comment (heejaechang) [inline src/Features/CSharp/Portable/CodeLens/CSharpCodeLensDisplayInfoService.cs]: you have symbol for methodNode. can it just follow up containing symbol to get name? or use ToDisplayString with full namestyle?

calling DeclaredSymbol seems unnecessary.

File: src/Features/CSharp/Portable/CodeLens/CSharpCodeLensDisplayInfoService.cs (hunk 1 above)
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/CodeLens/CSharpCodeLensDisplayInfoService.cs]: This looks like it's just building the metadata name.  SymbolDisplay should be able to handle this IIRC.

File: src/Features/Core/Portable/CodeLens/ICodeLensDisplayInfoService.cs [hunk 3]
```diff
@@ -15,5 +15,10 @@ internal interface ICodeLensDisplayInfoService : ILanguageService
         /// Gets the DisplayName for the given node
         /// </summary>
         string GetDisplayName(SemanticModel semanticModel, SyntaxNode node);
+
+        /// <summary>
+        /// Given a semantic model and methodNode, this method returns FQN with '+' as delimiter for the nested type
+        /// </summary>
+        string ConstructFullName(SemanticModel semanticModel, SyntaxNode methodNode);
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/CodeLens/ICodeLensDisplayInfoService.cs]: I think this shouldn't be necessary.

**Comment by panopticoncentral** [comment]
@CyrusNajmabadi I haven't walked all the way through the code, but given a method, they find all the methods that refer to it and then check if they are unit tests. I think they just check direct calls, they don't walk the call graph as far as I know.

File: src/Features/Core/Portable/CodeLens/ReferenceMethodDescriptor.cs [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+        /// <summary>
+        ///  Returns method's fully quilified name without parameters
+        /// </summary>
+        public string FullName { get; private set; }
+
+        /// <summary>
+        /// Returns method's file path.
+        /// </summary>
+        public string FullPath { get; private set; }
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/CodeLens/ReferenceMethodDescriptor.cs]: shoudl be called FilePath, not FullPath

File: src/Features/VisualBasic/Portable/CodeLens/VisualBasicDisplayInfoService.vb [hunk 5]
```diff
@@ -144,5 +144,54 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.CodeLens
...
+
+            Return rootName
+        End Function
+
+        Public Function ConstructFullName(semanticModel As SemanticModel, methodNode As SyntaxNode) As String Implements ICodeLensDisplayInfoService.ConstructFullName
+            Dim isTopLevelNodeAClass = True
+            Dim parent = methodNode.Parent
+            Dim name = semanticModel.GetDeclaredSymbol(methodNode).Name
+
+            While parent IsNot Nothing
```
Comment (CyrusNajmabadi) [inline src/Features/VisualBasic/Portable/CodeLens/VisualBasicDisplayInfoService.vb]: any time we're manually constructing a name, that is a bad smell.  We created SymbolDisplay precisely for the purpose of having a single unified location for building names.

File: src/Features/Core/Portable/CodeLens/CodeLenseReferenceService.cs [hunk 6]
```diff
@@ -167,43 +178,95 @@ private static SyntaxNode GetEnclosingCodeElementNode(Document document, SyntaxT
...
+                {
+                    var referenceTasks = progress.Locations
+                        .Where(location => location.Kind != LocationKind.MetadataFile && location.Kind != LocationKind.None)
+                        .Distinct(LocationComparer.Instance)
+                        .Select(
+                            location =>
+                                GetDescriptorOfEnclosingSymbolAsync(solution, location, cancellationToken))
+                        .ToArray();
+
+                    await Task.WhenAll(referenceTasks).ConfigureAwait(false);
```
Comment (heejaechang) [inline src/Features/Core/Portable/CodeLens/CodeLenseReferenceService.cs]: cancellation token?

File: src/Features/Core/Portable/CodeLens/CodeLenseReferenceService.cs [hunk 7]
```diff
@@ -167,43 +178,95 @@ private static SyntaxNode GetEnclosingCodeElementNode(Document document, SyntaxT
...
+
+                    return referenceTasks.Select(task => task.Result);
+                }, onCapped: null, searchCap: 0, cancellationToken: cancellationToken).ConfigureAwait(false);
+        }
+
+        private static SyntaxNode FindEnclosingMethodNode(Document document, int position)
+        {
+            var getSyntaxRootTask = document.GetSyntaxRootAsync();
+            var root = getSyntaxRootTask.Result;
```
Comment (heejaechang) [inline src/Features/Core/Portable/CodeLens/CodeLenseReferenceService.cs]: why not using async method?

File: src/Features/Core/Portable/CodeLens/CodeLenseReferenceService.cs [hunk 8]
```diff
@@ -167,43 +178,95 @@ private static SyntaxNode GetEnclosingCodeElementNode(Document document, SyntaxT
...
+            if (!root.FullSpan.Contains(position))
             {
                 return null;
             }
 
+            var token = root.FindToken(position);
+            var node = token.Parent;
+            var syntaxFacts = document.GetLanguageService<ISyntaxFactsService>();
+
+            while (node != null)
```
Comment (heejaechang) [inline src/Features/Core/Portable/CodeLens/CodeLenseReferenceService.cs]: syntax fact might already have this method. GetEnclosingMethod also, I think you can do node.GetAncestors( ... ).FirstOrDefault

File: src/Features/Core/Portable/CodeLens/CodeLenseReferenceService.cs [hunk 9]
```diff
@@ -167,43 +178,95 @@ private static SyntaxNode GetEnclosingCodeElementNode(Document document, SyntaxT
...
+                async progress =>
+                {
+                    var descriptorTasks =
+                        progress.Locations
+                        .Where(location => location.Kind != LocationKind.MetadataFile && location.Kind != LocationKind.None)
+                        .Select(location =>
+                            TryCreateMethodDescriptorAsync(location, solution, cancellationToken))
+                        .ToArray();
+
+                    await Task.WhenAll(descriptorTasks).ConfigureAwait(false);
```
Comment (heejaechang) [inline src/Features/Core/Portable/CodeLens/CodeLenseReferenceService.cs]: cancellation token?

File: src/Features/Core/Portable/CodeLens/CodeLenseReferenceService.cs [hunk 10]
```diff
@@ -18,7 +18,9 @@ namespace Microsoft.CodeAnalysis.CodeLens
     [ExportWorkspaceService(typeof(ICodeLensReferencesService)), Shared]
     internal sealed class CodeLensReferenceService : ICodeLensReferencesService
     {
-        public async Task<ReferenceCount?> GetReferenceCountAsync(Solution solution, DocumentId documentId, SyntaxNode syntaxNode, int maxSearchResults, CancellationToken cancellationToken)
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/CodeLens/CodeLenseReferenceService.cs]: Note: this filename is misspelled.  "CodeLensE"

File: src/Features/VisualBasic/Portable/CodeLens/VisualBasicDisplayInfoService.vb (hunk 5 above)
Comment (heejaechang) [inline src/Features/VisualBasic/Portable/CodeLens/VisualBasicDisplayInfoService.vb]: same as C# version

File: src/Features/Core/Portable/CodeLens/CodeLenseReferenceService.cs (hunk 9 above)
Comment (panopticoncentral) [inline src/Features/Core/Portable/CodeLens/CodeLenseReferenceService.cs]: WhenAll doesn't have an overload that takes a cancellation token, I guess it relies on the sub-tasks cancellation tokens?

---
In reply to: [74627996](https://github.com/dotnet/roslyn/pull/13109#discussion_r74627996) [](ancestors = 74627996)

File: src/Features/Core/Portable/CodeLens/CodeLenseReferenceService.cs [hunk 11]
```diff
@@ -167,43 +178,95 @@ private static SyntaxNode GetEnclosingCodeElementNode(Document document, SyntaxT
...
+
+                    await Task.WhenAll(referenceTasks).ConfigureAwait(false);
+
+                    return referenceTasks.Select(task => task.Result);
+                }, onCapped: null, searchCap: 0, cancellationToken: cancellationToken).ConfigureAwait(false);
+        }
+
+        private static SyntaxNode FindEnclosingMethodNode(Document document, int position)
+        {
+            var getSyntaxRootTask = document.GetSyntaxRootAsync();
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/CodeLens/CodeLenseReferenceService.cs]: GetSyntaxRootAsync.Result can be both very expensive and blocking.  This should be await'ed and should have a cancellation token passed to it.

File: src/Features/Core/Portable/CodeLens/CodeLenseReferenceService.cs (hunk 8 above)
Comment (panopticoncentral) [inline src/Features/Core/Portable/CodeLens/CodeLenseReferenceService.cs]: I found something like this in the find references code (GetEnclosingMethodOrPropertyOrField). It's not public and I don't need all the functionality, but it looks simpler, so I'll use similar code.

---
In reply to: [74627932](https://github.com/dotnet/roslyn/pull/13109#discussion_r74627932) [](ancestors = 74627932)

**Comment by panopticoncentral** [comment]
@CyrusNajmabadi The reason for manually constructing the name appears to be so that it matches the FQN format used elsewhere in VS--apparently they do FQNs with "+" between class names instead of "." (So it would be like "NS1.NS2.C1+C2+M1").

However, since the +/. distinction shouldn't really semantically important, I think we can try dropping this from the Roslyn side of things and have the VS side adapt. We'll see...

**Comment by CyrusNajmabadi** [comment]
The +  between nested types is the Metadata name for the type.  That should be something that SymbolDisplay supports natively IIRC.

**Comment by CyrusNajmabadi** [comment]
Looks lke i'm incorrect:
```
            if (format.TypeQualificationStyle == SymbolDisplayTypeQualificationStyle.NameAndContainingTypes ||
                format.TypeQualificationStyle == SymbolDisplayTypeQualificationStyle.NameAndContainingTypesAndNamespaces)
            {
                if (IncludeNamedType(symbol.ContainingType))
                {
                    symbol.ContainingType.Accept(this.NotFirstVisitor);
                    AddPunctuation(SyntaxKind.DotToken);
                }
            }
```

However, the right thing to do here would be to add an option to SymbolDisplay to ask for the metadata form of nested types, and to then call that.  

We really shouldn't be manually generating names *Ever*.

**Comment by panopticoncentral** [comment]
Yeah, I thought about adding an option to SymbolDisplay, but I think I'll take a look and see if VS can be adapted first since it's a fairly central public API addition for a relatively peripheral use.

**Comment by CyrusNajmabadi** [comment]
that's fine with me as well :)

**Comment by balajikris** [comment]
@panopticoncentral: Nice! Though, our team doesn't own this feature. Some other team does. If they too need an out of proc workspace, I guess it makes sense to move it too.

**Pull request merged**